bundles/org.openhab.binding.enphase/src/main/java/org/openhab/binding/enphase/internal/handler/EnvoyBridgeHandler.java

```diff
@@ -0,0 +1,409 @@
+    }
+
+    private static final long RETRY_RECONNECT_SECONDS = 10;
+
+    private final Logger logger = LoggerFactory.getLogger(EnvoyBridgeHandler.class);
+    private final EnvoyConnector connector;
+    private final EnvoyHostAddressCache envoyHostnameCache;
+
+    private @NonNullByDefault({}) EnvoyConfiguration configuration;
+    private @Nullable ScheduledFuture<?> updataDataFuture;
+    private @Nullable ScheduledFuture<?> updateHostnameFuture;
+    private @Nullable ExpiringCache<Map<String, @Nullable InverterDTO>> invertersCache;
+    private @Nullable ExpiringCache<Map<String, @Nullable DeviceDTO>> devicesCache;
+    private @Nullable EnvoyEnergyDTO productionDTO;
+    private @Nullable EnvoyEnergyDTO consumptionDTO;
+    private FeatureStatus consumptionSupported = FeatureStatus.UNKNOWN;
+    private FeatureStatus jsonSupported = FeatureStatus.UNKNOWN;
+
+    public EnvoyBridgeHandler(final Bridge thing, final HttpClient httpClient,
+            final EnvoyHostAddressCache envoyHostAddressCache) {
+        super(thing);
+        connector = new EnvoyConnector(httpClient);
+        this.envoyHostnameCache = envoyHostAddressCache;
+    }
+
+    @Override
+    public void handleCommand(final ChannelUID channelUID, final Command command) {
+        if (command instanceof RefreshType) {
+            refresh(channelUID);
+        }
+    }
+
+    private void refresh(final ChannelUID channelUID) {
+        final EnvoyEnergyDTO data = ENVOY_CHANNELGROUP_CONSUMPTION.equals(channelUID.getGroupId()) ? consumptionDTO
+                : productionDTO;
+
+        if (data == null) {
+            updateState(channelUID, UnDefType.UNDEF);
+        } else {
+            switch (channelUID.getIdWithoutGroup()) {
+                case ENVOY_WATT_HOURS_TODAY:
+                    updateState(channelUID, new QuantityType<>(data.wattHoursToday, Units.WATT_HOUR));
+                    break;
+                case ENVOY_WATT_HOURS_SEVEN_DAYS:
+                    updateState(channelUID, new QuantityType<>(data.wattHoursSevenDays, Units.WATT_HOUR));
+                    break;
+                case ENVOY_WATT_HOURS_LIFETIME:
+                    updateState(channelUID, new QuantityType<>(data.wattHoursLifetime, Units.WATT_HOUR));
+                    break;
+                case ENVOY_WATTS_NOW:
+                    updateState(channelUID, new QuantityType<>(data.wattsNow, Units.WATT));
+                    break;
+            }
+        }
+    }
+
+    @Override
+    public Collection<Class<? extends ThingHandlerService>> getServices() {
+        return Collections.singleton(EnphaseDevicesDiscoveryService.class);
+    }
+
+    @Override
+    public void initialize() {
+        configuration = getConfigAs(EnvoyConfiguration.class);
+        if (!EnphaseBindingConstants.isValidSerial(configuration.serialNumber)) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "Serial Number is not valid");
+        }
```

Comment:
Shouldn't you return immediately at this point? You can't complete initialization of the bridge if the config is invalid right?

Reply:
yes it should :smile: